File: demo-android-scala/src/main/scala/com/cengallut/textualapp/Demo1.scala
package com.cengallut.textualapp

import android.app.Activity
import android.os.Bundle
import com.cengallut.textual.core.CharGrid
import com.cengallut.textual.decoration.{Fill, Border}
import com.cengallut.textual.{TextualActivity, Action, TextualView}

/** Demonstration of decoration and touch event handling for a single Grid. */
class Demo1 extends TextualActivity {

  val decos = Seq(
    Fill.using('O'),
    Border.simple('+'))

  override def onBufferReady(buffer: CharGrid): Unit = {
    decos.foreach { _.decorate(buffer) }
    buffer.notifyChanged()

    val touchListener = Action(buffer).touch { (x, y) =>
      buffer.charAt(x, y) match {
        case 'O' => buffer.setChar(x, y, 'X')
        case _   => buffer.setChar(x, y, 'O')
      }
      buffer.notifyChanged()
    }

    view.setOnTouchListener(touchListener)

  }

}

File: demo-android-java/src/main/java/com/cengallut/textualjavademo/Demo1.java
package com.cengallut.textualjavademo;

import com.cengallut.textual.Action;
import com.cengallut.textual.TextualActivity;
import com.cengallut.textual.core.CharGrid;
import com.cengallut.textual.core.GridAgent;
import com.cengallut.textual.decoration.Border;
import com.cengallut.textual.decoration.Decoration;
import com.cengallut.textual.decoration.Fill;

import java.util.Arrays;
import java.util.List;

/** Demonstration of decoration and touch event handling for a single Grid. */
public class Demo1 extends TextualActivity
        implements GridAgent {

    final List<Decoration> decos = Arrays.asList(
            Fill.using('o'),
            Border.simple('#'));

    CharGrid buffer = null;

    @Override
    public void onBufferReady(CharGrid b) {
        buffer = b;

        for (Decoration d : decos) {
            d.decorate(b);
        }

        view().setOnTouchListener(new Action(b).touch(this));
        buffer.notifyChanged();
    }

    @Override
    public void onAction(int x, int y) {
        if (buffer.charAt(x, y) == 'O') {
            buffer.setChar(x, y, 'X');
        } else {
            buffer.setChar(x, y, 'O');
        }
        buffer.notifyChanged();
    }
}
